diff -rupN src-orig/Makefile.in src/Makefile.in
--- src-orig/Makefile.in	2014-02-23 20:23:25.245625275 -0800
+++ src/Makefile.in	2014-02-23 20:24:21.613626013 -0800
@@ -55,7 +55,7 @@ bin_PROGRAMS = groups$(EXEEXT) login$(EX
 sbin_PROGRAMS = nologin$(EXEEXT)
 ubin_PROGRAMS = faillog$(EXEEXT) lastlog$(EXEEXT) chage$(EXEEXT) \
 	chfn$(EXEEXT) chsh$(EXEEXT) expiry$(EXEEXT) gpasswd$(EXEEXT) \
-	newgrp$(EXEEXT) passwd$(EXEEXT)
+	newgrp$(EXEEXT) passwd$(EXEEXT) newgidmap$(EXEEXT) newuidmap$(EXEEXT)
 usbin_PROGRAMS = chgpasswd$(EXEEXT) chpasswd$(EXEEXT) \
 	groupadd$(EXEEXT) groupdel$(EXEEXT) groupmems$(EXEEXT) \
 	groupmod$(EXEEXT) grpck$(EXEEXT) grpconv$(EXEEXT) \
