diff -rupN src-orig/Makefile.in src/Makefile.in
--- src-orig/Makefile.in	2014-02-23 19:39:33.153590789 -0800
+++ src/Makefile.in	2014-02-23 19:52:48.809601214 -0800
@@ -181,10 +181,18 @@ newgrp_SOURCES = newgrp.c
 newgrp_OBJECTS = newgrp.$(OBJEXT)
 newgrp_DEPENDENCIES = $(am__DEPENDENCIES_2) $(am__DEPENDENCIES_1) \
 	$(am__DEPENDENCIES_1)
+newgidmap_SOURCES = newgidmap.c
+newgidmap_OBJECTS = newgidmap.$(OBJEXT)
+newgidmap_DEPENDENCIES = $(am__DEPENDENCIES_2) $(am__DEPENDENCIES_1) \
+	$(am__DEPENDENCIES_1)
 newusers_SOURCES = newusers.c
 newusers_OBJECTS = newusers.$(OBJEXT)
 newusers_DEPENDENCIES = $(am__DEPENDENCIES_2) $(am__DEPENDENCIES_1) \
 	$(am__DEPENDENCIES_1) $(am__DEPENDENCIES_1)
+newuidmap_SOURCES = newuidmap.c
+newuidmap_OBJECTS = newuidmap.$(OBJEXT)
+newuidmap_DEPENDENCIES = $(am__DEPENDENCIES_2) $(am__DEPENDENCIES_1) \
+	$(am__DEPENDENCIES_1) $(am__DEPENDENCIES_1)
 nologin_SOURCES = nologin.c
 nologin_OBJECTS = nologin.$(OBJEXT)
 nologin_DEPENDENCIES =
@@ -248,13 +256,13 @@ SOURCES = chage.c chfn.c chgpasswd.c chp
 	groupmod.c groups.c grpck.c grpconv.c grpunconv.c id.c \
 	lastlog.c $(login_SOURCES) logoutd.c newgrp.c newusers.c \
 	nologin.c passwd.c pwck.c pwconv.c pwunconv.c $(su_SOURCES) \
-	sulogin.c useradd.c userdel.c usermod.c vipw.c
+	sulogin.c useradd.c userdel.c usermod.c vipw.c newgidmap.c newuidmap.c
 DIST_SOURCES = chage.c chfn.c chgpasswd.c chpasswd.c chsh.c expiry.c \
 	faillog.c gpasswd.c groupadd.c groupdel.c groupmems.c \
 	groupmod.c groups.c grpck.c grpconv.c grpunconv.c id.c \
 	lastlog.c $(login_SOURCES) logoutd.c newgrp.c newusers.c \
 	nologin.c passwd.c pwck.c pwconv.c pwunconv.c $(su_SOURCES) \
-	sulogin.c useradd.c userdel.c usermod.c vipw.c
+	sulogin.c useradd.c userdel.c usermod.c vipw.c newgidmap.c newuidmap.c
 am__can_run_installinfo = \
   case $$AM_UPDATE_INFO_DIR in \
     n|no|NO) false;; \
@@ -423,7 +431,7 @@ INCLUDES = \
 	-I$(top_srcdir)/libmisc
 
 suidbins = su
-suidubins = chage chfn chsh expiry gpasswd newgrp passwd
+suidubins = chage chfn chsh expiry gpasswd newgrp passwd newgidmap newuidmap
 @WITH_TCB_TRUE@shadowsgidubins = passwd
 LDADD = $(INTLLIBS) \
 		 $(LIBTCB) \
@@ -454,7 +462,9 @@ login_SOURCES = \
 
 login_LDADD = $(LDADD) $(LIBPAM) $(LIBAUDIT) $(LIBCRYPT_NOPAM) $(LIBSKEY) $(LIBMD)
 newgrp_LDADD = $(LDADD) $(LIBAUDIT) $(LIBCRYPT)
+newgidmap_LDADD = $(LDADD) $(LIBAUDIT) $(LIBCRYPT)
 newusers_LDADD = $(LDADD) $(LIBPAM) $(LIBSELINUX) $(LIBCRYPT)
+newuidmap_LDADD = $(LDADD) $(LIBPAM) $(LIBSELINUX) $(LIBCRYPT)
 nologin_LDADD = 
 passwd_LDADD = $(LDADD) $(LIBPAM) $(LIBCRACK) $(LIBAUDIT) $(LIBSELINUX) $(LIBCRYPT_NOPAM)
 pwck_LDADD = $(LDADD) $(LIBSELINUX)
@@ -760,9 +770,15 @@ logoutd$(EXEEXT): $(logoutd_OBJECTS) $(l
 newgrp$(EXEEXT): $(newgrp_OBJECTS) $(newgrp_DEPENDENCIES) $(EXTRA_newgrp_DEPENDENCIES) 
 	@rm -f newgrp$(EXEEXT)
 	$(LINK) $(newgrp_OBJECTS) $(newgrp_LDADD) $(LIBS)
-newusers$(EXEEXT): $(newusers_OBJECTS) $(newusers_DEPENDENCIES) $(EXTRA_newusers_DEPENDENCIES) 
+newgidmap$(EXEEXT): $(newgidmap_OBJECTS) $(newgidmap_DEPENDENCIES) $(EXTRA_newgidmap_DEPENDENCIES)
+	@rm -f newgidmap$(EXEEXT)
+	$(LINK) $(newgidmap_OBJECTS) $(newgidmap_LDADD) $(LIBS)
+newusers$(EXEEXT): $(newusers_OBJECTS) $(newusers_DEPENDENCIES) $(EXTRA_newusers_DEPENDENCIES)
 	@rm -f newusers$(EXEEXT)
 	$(LINK) $(newusers_OBJECTS) $(newusers_LDADD) $(LIBS)
+newuidmap$(EXEEXT): $(newuidmap_OBJECTS) $(newuidmap_DEPENDENCIES) $(EXTRA_newuidmap_DEPENDENCIES)
+	@rm -f newuidmap$(EXEEXT)
+	$(LINK) $(newuidmap_OBJECTS) $(newuidmap_LDADD) $(LIBS)
 nologin$(EXEEXT): $(nologin_OBJECTS) $(nologin_DEPENDENCIES) $(EXTRA_nologin_DEPENDENCIES) 
 	@rm -f nologin$(EXEEXT)
 	$(LINK) $(nologin_OBJECTS) $(nologin_LDADD) $(LIBS)
@@ -825,7 +841,9 @@ distclean-compile:
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/login_nopam.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/logoutd.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/newgrp.Po@am__quote@
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/newgidmap.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/newusers.Po@am__quote@
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/newuidmap.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/nologin.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/passwd.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/pwck.Po@am__quote@
@@ -1075,7 +1093,7 @@ uninstall-am: uninstall-binPROGRAMS unin
 	uninstall-sbinPROGRAMS uninstall-ubinPROGRAMS \
 	uninstall-usbinPROGRAMS
 
-@ACCT_TOOLS_SETUID_TRUE@	suidubins += chage chgpasswd chpasswd groupadd groupdel groupmod newusers useradd userdel usermod
+@ACCT_TOOLS_SETUID_TRUE@	suidubins += chage chgpasswd chpasswd groupadd groupdel groupmod newusers useradd userdel usermod newuidmap newgidmap
 
 @WITH_TCB_TRUE@suidubins -= passwd
 
