diff -rupN shadow-4.1.5.1-orig/lib/Makefile shadow-4.1.5.1/lib/Makefile
--- shadow-4.1.5.1-orig/lib/Makefile	2014-02-23 17:54:44.189508390 -0800
+++ shadow-4.1.5.1/lib/Makefile	2014-02-23 17:54:57.865508569 -0800
@@ -72,15 +72,15 @@ am__libshadow_la_SOURCES_DIST = commonio
 	nscd.h pam_defs.h port.c port.h prototypes.h pwauth.c pwauth.h \
 	pwio.c pwio.h pwmem.c selinux.c semanage.c sgetgrent.c \
 	sgetpwent.c sgetspent.c sgroupio.c sgroupio.h shadow.c \
-	shadowio.c shadowio.h shadowmem.c spawn.c utent.c tcbfuncs.c \
-	tcbfuncs.h
+	shadowio.c shadowio.h shadowmem.c spawn.c subordinateio.c utent.c \
+	tcbfuncs.c tcbfuncs.h
 #am__objects_1 = tcbfuncs.lo
 am_libshadow_la_OBJECTS = commonio.lo encrypt.lo fields.lo fputsx.lo \
 	getdef.lo get_gid.lo getlong.lo get_pid.lo get_uid.lo \
 	getulong.lo groupio.lo groupmem.lo gshadow.lo lockpw.lo \
 	nscd.lo port.lo pwauth.lo pwio.lo pwmem.lo selinux.lo \
 	semanage.lo sgetgrent.lo sgetpwent.lo sgetspent.lo sgroupio.lo \
-	shadow.lo shadowio.lo shadowmem.lo spawn.lo utent.lo \
+	shadow.lo shadowio.lo shadowmem.lo spawn.lo subordinateio.lo utent.lo \
 	$(am__objects_1)
 libshadow_la_OBJECTS = $(am_libshadow_la_OBJECTS)
 libshadow_la_LINK = $(LIBTOOL) --tag=CC $(AM_LIBTOOLFLAGS) \
@@ -267,7 +267,7 @@ libshadow_la_SOURCES = commonio.c common
 	pam_defs.h port.c port.h prototypes.h pwauth.c pwauth.h pwio.c \
 	pwio.h pwmem.c selinux.c semanage.c sgetgrent.c sgetpwent.c \
 	sgetspent.c sgroupio.c sgroupio.h shadow.c shadowio.c \
-	shadowio.h shadowmem.c spawn.c utent.c $(am__append_1)
+	shadowio.h shadowmem.c spawn.c subordinateio.c utent.c $(am__append_1)
 
 # These files are unneeded for some reason, listed in
 # order of appearance:
@@ -329,58 +329,59 @@ mostlyclean-compile:
 distclean-compile:
 	-rm -f *.tab.c
 
-#include ./$(DEPDIR)/commonio.Plo
-#include ./$(DEPDIR)/encrypt.Plo
-#include ./$(DEPDIR)/fields.Plo
-#include ./$(DEPDIR)/fputsx.Plo
-#include ./$(DEPDIR)/get_gid.Plo
-#include ./$(DEPDIR)/get_pid.Plo
-#include ./$(DEPDIR)/get_uid.Plo
-#include ./$(DEPDIR)/getdef.Plo
-#include ./$(DEPDIR)/getlong.Plo
-#include ./$(DEPDIR)/getulong.Plo
-#include ./$(DEPDIR)/groupio.Plo
-#include ./$(DEPDIR)/groupmem.Plo
-#include ./$(DEPDIR)/gshadow.Plo
-#include ./$(DEPDIR)/lockpw.Plo
-#include ./$(DEPDIR)/nscd.Plo
-#include ./$(DEPDIR)/port.Plo
-#include ./$(DEPDIR)/pwauth.Plo
-#include ./$(DEPDIR)/pwio.Plo
-#include ./$(DEPDIR)/pwmem.Plo
-#include ./$(DEPDIR)/selinux.Plo
-#include ./$(DEPDIR)/semanage.Plo
-#include ./$(DEPDIR)/sgetgrent.Plo
-#include ./$(DEPDIR)/sgetpwent.Plo
-#include ./$(DEPDIR)/sgetspent.Plo
-#include ./$(DEPDIR)/sgroupio.Plo
-#include ./$(DEPDIR)/shadow.Plo
-#include ./$(DEPDIR)/shadowio.Plo
-#include ./$(DEPDIR)/shadowmem.Plo
-#include ./$(DEPDIR)/spawn.Plo
-#include ./$(DEPDIR)/tcbfuncs.Plo
-#include ./$(DEPDIR)/utent.Plo
+include ./$(DEPDIR)/commonio.Plo
+include ./$(DEPDIR)/encrypt.Plo
+include ./$(DEPDIR)/fields.Plo
+include ./$(DEPDIR)/fputsx.Plo
+include ./$(DEPDIR)/get_gid.Plo
+include ./$(DEPDIR)/get_pid.Plo
+include ./$(DEPDIR)/get_uid.Plo
+include ./$(DEPDIR)/getdef.Plo
+include ./$(DEPDIR)/getlong.Plo
+include ./$(DEPDIR)/getulong.Plo
+include ./$(DEPDIR)/groupio.Plo
+include ./$(DEPDIR)/groupmem.Plo
+include ./$(DEPDIR)/gshadow.Plo
+include ./$(DEPDIR)/lockpw.Plo
+include ./$(DEPDIR)/nscd.Plo
+include ./$(DEPDIR)/port.Plo
+include ./$(DEPDIR)/pwauth.Plo
+include ./$(DEPDIR)/pwio.Plo
+include ./$(DEPDIR)/pwmem.Plo
+include ./$(DEPDIR)/selinux.Plo
+include ./$(DEPDIR)/semanage.Plo
+include ./$(DEPDIR)/sgetgrent.Plo
+include ./$(DEPDIR)/sgetpwent.Plo
+include ./$(DEPDIR)/sgetspent.Plo
+include ./$(DEPDIR)/sgroupio.Plo
+include ./$(DEPDIR)/shadow.Plo
+include ./$(DEPDIR)/shadowio.Plo
+include ./$(DEPDIR)/shadowmem.Plo
+include ./$(DEPDIR)/spawn.Plo
+include ./$(DEPDIR)/subordiateio.Plo
+include ./$(DEPDIR)/tcbfuncs.Plo
+include ./$(DEPDIR)/utent.Plo
 
 .c.o:
-#	$(COMPILE) -MT $@ -MD -MP -MF $(DEPDIR)/$*.Tpo -c -o $@ $<
-#	$(am__mv) $(DEPDIR)/$*.Tpo $(DEPDIR)/$*.Po
-#	source='$<' object='$@' libtool=no 
-#	DEPDIR=$(DEPDIR) $(CCDEPMODE) $(depcomp) 
-	$(COMPILE) -c $<
+	$(COMPILE) -MT $@ -MD -MP -MF $(DEPDIR)/$*.Tpo -c -o $@ $<
+	$(am__mv) $(DEPDIR)/$*.Tpo $(DEPDIR)/$*.Po
+#	source='$<' object='$@' libtool=no \
+#	DEPDIR=$(DEPDIR) $(CCDEPMODE) $(depcomp) \
+#	$(COMPILE) -c $<
 
 .c.obj:
-#	$(COMPILE) -MT $@ -MD -MP -MF $(DEPDIR)/$*.Tpo -c -o $@ `$(CYGPATH_W) '$<'`
-#	$(am__mv) $(DEPDIR)/$*.Tpo $(DEPDIR)/$*.Po
-#	source='$<' object='$@' libtool=no 
-#	DEPDIR=$(DEPDIR) $(CCDEPMODE) $(depcomp) 
-	$(COMPILE) -c `$(CYGPATH_W) '$<'`
+	$(COMPILE) -MT $@ -MD -MP -MF $(DEPDIR)/$*.Tpo -c -o $@ `$(CYGPATH_W) '$<'`
+	$(am__mv) $(DEPDIR)/$*.Tpo $(DEPDIR)/$*.Po
+#	source='$<' object='$@' libtool=no \
+#	DEPDIR=$(DEPDIR) $(CCDEPMODE) $(depcomp) \
+#	$(COMPILE) -c `$(CYGPATH_W) '$<'`
 
 .c.lo:
-#	$(LTCOMPILE) -MT $@ -MD -MP -MF $(DEPDIR)/$*.Tpo -c -o $@ $<
-#	$(am__mv) $(DEPDIR)/$*.Tpo $(DEPDIR)/$*.Plo
-#	source='$<' object='$@' libtool=yes 
-#	DEPDIR=$(DEPDIR) $(CCDEPMODE) $(depcomp) 
-	$(LTCOMPILE) -c -o $@ $<
+	$(LTCOMPILE) -MT $@ -MD -MP -MF $(DEPDIR)/$*.Tpo -c -o $@ $<
+	$(am__mv) $(DEPDIR)/$*.Tpo $(DEPDIR)/$*.Plo
+#	source='$<' object='$@' libtool=yes \
+#	DEPDIR=$(DEPDIR) $(CCDEPMODE) $(depcomp) \
+#	$(LTCOMPILE) -c -o $@ $<
 
 mostlyclean-libtool:
 	-rm -f *.lo
