diff -rupN docker-0.8.1-orig/graph.go docker-0.8.1/graph.go
--- docker-0.8.1-orig/graph.go	2014-02-24 17:33:00.421019960 -0800
+++ docker-0.8.1/graph.go	2014-02-24 18:03:18.003043774 -0800
@@ -37,6 +37,9 @@ func NewGraph(root string, driver graphd
 		return nil, err
 	}
 
+	// HACK: hard coding a user namespace uid,gid
+	os.Chown(root, 100000, 100000)
+
 	graph := &Graph{
 		Root:    abspath,
 		idIndex: utils.NewTruncIndex(),
